subprojects/kotlin-dsl-provider-plugins/src/main/kotlin/org/gradle/kotlin/dsl/provider/plugins/DefaultProjectSchemaProvider.kt

```diff
@@ -57,26 +58,36 @@ class DefaultProjectSchemaProvider : ProjectSchemaProvider {
 
 internal
 data class TargetTypedSchema(
-    val extensions: List<ProjectSchemaEntry<TypeOf<*>>>,
-    val conventions: List<ProjectSchemaEntry<TypeOf<*>>>,
-    val tasks: List<ProjectSchemaEntry<TypeOf<*>>>,
-    val containerElements: List<ProjectSchemaEntry<TypeOf<*>>>
+    val extensions: Set<ProjectSchemaEntry<TypeOf<*>>>,
+    val conventions: Set<ProjectSchemaEntry<TypeOf<*>>>,
+    val tasks: Set<ProjectSchemaEntry<TypeOf<*>>>,
+    val containerElements: Set<ProjectSchemaEntry<TypeOf<*>>>
 )
 
 
 internal
 fun targetSchemaFor(target: Any, targetType: TypeOf<*>): TargetTypedSchema {
 
-    val extensions = mutableListOf<ProjectSchemaEntry<TypeOf<*>>>()
-    val conventions = mutableListOf<ProjectSchemaEntry<TypeOf<*>>>()
-    val tasks = mutableListOf<ProjectSchemaEntry<TypeOf<*>>>()
-    val containerElements = mutableListOf<ProjectSchemaEntry<TypeOf<*>>>()
+    val extensions = mutableSetOf<ProjectSchemaEntry<TypeOf<*>>>()
+    val conventions = mutableSetOf<ProjectSchemaEntry<TypeOf<*>>>()
+    val tasks = mutableSetOf<ProjectSchemaEntry<TypeOf<*>>>()
+    val containerElements = mutableSetOf<ProjectSchemaEntry<TypeOf<*>>>()
```

Comment:
:x: I think it would be preferable to keep lists here and make the recursive function below skip already seen targets instead. This would avoid unnecessary processing. The current change adds overhead.

Reply:
I don't understand how the recursive algorithm could do those checks (in a generic manner) without incurring the same overhead ...

Maybe checking for duplicates makes sense only for some, like `containerElements` and not for others, like `tasks`?

Maybe duplicates can arise also by elements being in multiple collections, which we don't check for even now?

Let's discuss live.